File: js/src/api/userAPI.ts
import request from '@/api/request'
import jwtDecode from "jwt-decode"

import getIdAndName from '@/common/getIdAndName'

export const userRegister =async (user: string)=> {
  const result = await request(`/user/register`, user, 'PUT')
  if(result.error) {
    return {
      isSuccess: false,
      message: result.error
    }
  } else if(!result.isSuccess) {
    return {
      isSuccess: false,
      message: result.message
    }
  } else if(result.isSuccess) {
    return {
      isSuccess: true,
      message: result.data
    }
  } else {
    return {
      isSuccess: false,
      message: "Unknow error!"
    }
  }
}

export const userLogin = async(user:string)=> {
  const result = await request(`/user/login`, user, 'POST')
  if(result.error) {
    return {
      isSuccess: false,
      message: result.error
    }
  } else if(!result.isSuccess) {
    return {
      isSuccess: false,
      message: result.message
    }
  } else if(result.isSuccess) {
    return {
      isSuccess: true,
      message: result.data
    }
  } else {
    return {
      isSuccess: false,
      message: "Unknow error!"
    }
  }
}

export const getUserById = async(targetId: number, userId: number)=> {
  const result = await request(`/user/id`, {targetId, userId})
  if(result.error) {
    return result.error
  } else if(result.isSuccess) {
    return result.data
  } else {
    return result.message
  }
}

export const checkToken = async()=> {
  const accessToken = localStorage.getItem("accessToken")
  const refreshToken = localStorage.getItem("refreshToken")
  let shouldRefresh = false
  let message = "登陆验证通过"

  if(!accessToken || !refreshToken) {
    //前端没有对应的token，说明用户未曾登录
    localStorage.removeItem("accessToken")
    return {
      isLogin: false,
      message: "登录信息不合法"
    }
  }

  let decode: any

  //解码accessToken，用于取得过期时间
  try {
    decode = jwtDecode(accessToken)
  } catch(e) {
    return {
      isLogin: false,
      message: "请登录后再执行操作"
    }
  }

  //当前时间戳（毫秒单位）
  const now = Date.now()

  //token过期的时间戳（秒单位）
  const tokenExp = decode.exp

  //剩余时间（分钟单位）
  const restTime = (tokenExp - Math.ceil(now/1000))/60

  //如果accessToken剩余时间小于30分钟，提醒后端需要刷新token
  if(restTime < 30) {
    shouldRefresh = true
  }

  //获取后端检查token的结果
  const result = await request(`/user/checkToken`, {}, "GET", {
    accessToken, refreshToken, shouldRefresh
  })

  //结果有accessToken，说明accessToken需要刷新
  if(result.data?.accessToken) {
    localStorage.setItem("accessToken", result.data.accessToken)
    //结果有refreshToken，说明refreshToken需要刷新
    if(result.data.refreshToken) {
      localStorage.setItem("refreshToken", result.data.refreshToken)
    }
  }

  //结果有message（错误信息）
  if(result.message) {
    message = result.message
  }  


  //返回登陆验证的状态，isSuccess
  return {isLogin:result.isSuccess, message: message}

}

export const auditorVerify = async()=> {
  const userId = getIdAndName().id
  const auditorId = localStorage.getItem("auditorId")

  console.log(auditorId)

  if(auditorId==null) {
    return {
      isSuccess: false,
      message: "You're not auditor"
    }
  }

  const result = await request(`/user/audit/verify`, {auditorId, userId}, "get", {
    token: localStorage.getItem("accessToken") || ""
  })
  if(result.error) {
    return {
      isSuccess: false,
      message: result.error
    }
  } else if(result.isSuccess) {
    return {
      isSuccess: true,
      message: result.data
    }
  } else {
    return {
      isSuccess: false,
      message: result.message
    }
  }
}

export const getUserContribute = async (contributorId: number, userId: number, pageNum: number = 1, pageSize: number = 15) => {
  const result = await request(`/user/contribute/all`, {contributorId, userId, pageNum, pageSize})
  if(result.error) {
    return result.error
  } else if(result.isSuccess) {
    return result.data
  } else {
    return result.message
  }
}

export const doThankContribute = async(area: string, historyId:number, userId: number)=> {
  const result = await request(`/user/contribute/doThank`, {area, historyId, userId})
  console.log(result)
  return result.isSuccess
}

export const undoThankContribute = async(area: string, historyId:number, userId: number)=> {
  const result = await request(`/user/contribute/undoThank`, {area, historyId, userId})
  console.log(result)
  return result.isSuccess
}

export const myFollows = async(userId: number, pageNum: number, pageSize: number)=> {
  const result = await request(`/user/follow/myFollows`, {userId, pageNum, pageSize})
  if(result.error) {
    return result.error
  } else if(result.isSuccess) {
    return result.data
  } else {
    return result.error
  }
}

export const myFollowers = async(userId: number, pageNum: number, pageSize: number)=> {
  const result = await request(`/user/follow/myFollowers`, {userId, pageNum, pageSize})
  if(result.error) {
    return result.error
  } else if(result.isSuccess) {
    return result.data
  } else {
    return result.error
  }
}

export const doFollow = async(target: number, userId: number, userName: string)=> {
  return await request(`/user/follow/do/${target}`, {userId, userName})
}

export const undoFollow = async(target: number, userId: number, userName: string)=> {
  return await request(`/user/follow/undo/${target}`, {userId, userName})
}

export const getAuditData = async(auditorId: number) => {
  return await request(`/user/audit/data`, {auditorId})
}

export const logout = async(userId: number)=> {
  return await request(`/user/logout/${userId}`, {}, "post")
}

export const update = async(userDO: any)=> {
  const str = JSON.stringify(userDO)
  return await request(`/user/update`, str, "post")
}

File: java/project/src/main/java/com/example/project/api/UserAPI.java
package com.example.project.api;

import com.auth0.jwt.interfaces.Claim;
import com.example.project.common.Paging;
import com.example.project.common.Result;
import com.example.project.dataobject.UserDO;
import com.example.project.model.User;
import com.example.project.service.UserService;
import com.example.project.util.JwtUtil;
import com.example.project.util.KeywordUtil;
import org.apache.commons.codec.digest.DigestUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.*;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpSession;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

/**
 * 与user相关的操作的api
 */
@Controller
@RequestMapping("/user")
public class UserAPI {

    @Autowired
    private RedisTemplate redisTemplate;

    @Autowired
    private UserService userService;

    @Scheduled(cron = "* */15 * * * ?")
    public void addContributePoint() {
        userService.addContributePoint();
    }

    @Scheduled(cron = "* */15 * * * ?")
    public void addFollowFan() {
        userService.addFollowFan();
        userService.addAuditPoint();
    }

    /**
     * 每天删除一次今日审核数据
     */
    @Scheduled(cron = "0 0 0 * * ?")
    public void delTodayAudit() {
        userService.clearTodayAudit();
    }

    /**
     * user注册
     * @param userDO
     * @param session
     * @return
     */
    @PutMapping("/register")
    @ResponseBody
    public Result register(@RequestBody UserDO userDO, HttpSession session) {

        return userService.register(userDO, session);
    }

    /**
     * user登录
     * @param userDO
     * @param session
     * @return
     */
    @PostMapping("/login")
    @ResponseBody
    public Result login(@RequestBody UserDO userDO, HttpSession session) {
        return userService.login(userDO, session);
    }

    @PostMapping("/logout/{userId}")
    @ResponseBody
    public Result logout(@PathVariable("userId")Long userId) {
        return userService.logout(userId);
    }

    @GetMapping("/id")
    @ResponseBody
    public Result getUserById(@RequestParam("targetId")Long targetId, @RequestParam("userId")Long userId) {
        User user = userService.getUserById(targetId, userId);
        if(user==null) {
            return Result.error("Get user by id failed!");
        } else {
            return Result.success(user);
        }
    }

    @PostMapping("/update")
    @ResponseBody
    public Result updateUser(@RequestBody UserDO userDO) {
        int num = userService.updateUser(userDO);
        if(num==1) {
            return Result.success("Update user info success!");
        } else {
            return Result.success("Update user info failed!");
        }
    }

    /**
     * 检查或更新user token
     * @param request
     * @return
     */
    @GetMapping("/checkToken")
    @ResponseBody
    public Result refreshToken(HttpServletRequest request) {
        String accessToken = request.getHeader("accessToken");
        String refreshToken = request.getHeader("refreshToken");
        String shouldRefresh = request.getHeader("shouldRefresh");

        Map<String, Claim> infoMap;

        //解析token，获取用户信息
        infoMap = JwtUtil.parseToken(accessToken);
        Long id = infoMap.get("id").asLong();
        String name = infoMap.get("name").asString();

        //token已失效
        try {
            JwtUtil.verifyUserToken(accessToken);
        } catch (Exception e) {
            redisTemplate.delete(JwtUtil.REFRESH_PREFIX+id);
            return Result.error("登陆状态已失效，请重新登录！");
        }

        //前端判断需要刷新token（前端持有的accessToken即将过期）
        if(shouldRefresh.equals("true")) {

            //在redis中取得refreshToken
            String checkRefreshToken = (String) redisTemplate.opsForValue().get(JwtUtil.REFRESH_PREFIX+id);

            //判断前后端的refreshToken是否可以对应
            if(refreshToken.equals(checkRefreshToken)) {

                Map resultMap = null;

                try {
                    //可以对应，获取新的accessToken
                    String newAccessToken = JwtUtil.createUserToken(id, name, JwtUtil.ACCESS_EXPIRE);

                    resultMap = new HashMap<>();
                    resultMap.put("accessToken", newAccessToken);

                    Long restTime = redisTemplate.getExpire(JwtUtil.REFRESH_PREFIX+id, TimeUnit.MINUTES);

                    //如果refreshToken也即将过期
                    if(restTime < 30) {
                        //生成新的refreshToken并存入redis
                        String newRefreshToken = JwtUtil.makeRefreshToken();
                        redisTemplate.opsForValue().set(JwtUtil.REFRESH_PREFIX+id, newRefreshToken);

                        resultMap.put("refreshToken", newRefreshToken);
                    }

                    //返回token
                    return Result.success(resultMap);

                } catch (Exception e) {
                    //出任何错误都标识为登陆状态失效
                    redisTemplate.delete(JwtUtil.REFRESH_PREFIX+id);
                    return Result.error("登陆状态已失效，请重新登录！");
                }

            } else {
                //前后端的refreshToken对应不上
                redisTemplate.delete(JwtUtil.REFRESH_PREFIX+id);
                return Result.error("登陆状态已失效，请重新登录！");
            }
        } else {
            return Result.success(null);
        }

    }

    /**
     * 检查audit资格
     * @param auditorId
     * @param userId
     * @return
     */
    @GetMapping("/audit/verify")
    @ResponseBody
    public Result auditVerify(@RequestParam("auditorId") Long auditorId,@RequestParam("userId") Long userId) {
        return userService.auditVerify(auditorId, userId);
    }

    @GetMapping("/audit/data")
    @ResponseBody
    public Result getAuditData(@RequestParam("auditorId") Long auditorId) {
        return userService.getAuditData(auditorId);
    }

    /**
     * 通过userId获取全部贡献
     * @param userId
     * @return
     */
    @GetMapping("/contribute/all")
    @ResponseBody
    public Result getUserContribute(@RequestParam(value="contributorId") Long contributorId,
                                    @RequestParam(value="userId") Long userId,
                                    @RequestParam(value="pageNum", defaultValue="1") int pageNum,
                                    @RequestParam(value="pageSize", defaultValue="15") int pageSize) {
        Paging result = userService.getUserContribute(contributorId, userId, pageNum, pageSize);
        if(result.getData()==null) {
            return Result.error("Get user contribute failed!");
        } else {
            return Result.success(result);
        }
    }


    /**
     * 对用户贡献记录进行点赞
     * @param area
     * @param historyId
     * @param userId
     * @return
     */
    @GetMapping("/contribute/doThank")
    @ResponseBody
    public Result doThankContribute(@RequestParam("area")String area, @RequestParam("historyId")Long historyId, @RequestParam("userId")Long userId) {
        long result = userService.doThankContribute(area, historyId, userId);
        if(result==1) {
            return Result.success("thank contribute success!");
        } else {
            return Result.error("repetitive thank!");
        }
    }

    /**
     * 对用户贡献记录进行取消感谢
     * @param area
     * @param historyId
     * @param userId
     * @return
     */
    @GetMapping("/contribute/undoThank")
    @ResponseBody
    public Result undoThankContribute(@RequestParam("area")String area, @RequestParam("historyId")Long historyId, @RequestParam("userId")Long userId) {
        long result = userService.undoThankContribute(area, historyId, userId);
        if(result==1) {
            return Result.success("cancel thank contribute success!");
        } else {
            return Result.error("repetitive cancel thank!");
        }
    }

    @GetMapping("/getPost")
    @ResponseBody
    public Result<Map> getPost(@RequestParam("area")String area, @RequestParam("historyId")Long historyId, @RequestParam("userId")Long userId) {
        return Result.success(userService.getPost(area, historyId, userId));
    }

    @GetMapping("/follow/do/{target}")
    @ResponseBody
    public Result follow(@PathVariable("target")Long target, @RequestParam("userId") Long userId, @RequestParam("userName") String userName) {
        User user = new User();
        user.setId(userId);
        user.setName(userName);
        if(userService.isFollow(target, userId)) {
            return Result.error("You're already follow "+target+"!");
        } else {
            userService.follow(target, user);
            return Result.success("Follow "+target+" success!");
        }
    }

    @GetMapping("/follow/undo/{target}")
    @ResponseBody
    public Result unFollow(@PathVariable("target")Long target, @RequestParam("userId") Long userId, @RequestParam("userName") String userName) {
        User user = new User();
        user.setId(userId);
        user.setName(userName);
        if(userService.isFollow(target, userId)) {
            userService.unFollow(target, user);
            return Result.success("Follow "+target+" cancel success!");
        } else {
            return Result.error("You're already unFollow "+target+"!");
        }
    }

    @GetMapping("/follow/myFollows")
    @ResponseBody
    public Result myFollows(@RequestParam("userId")Long userId, @RequestParam("pageNum") int pageNum, @RequestParam("pageSize") int pageSize) {
        return Result.success(userService.myFollows(userId, pageNum, pageSize));
    }

    @GetMapping("/follow/myFollowers")
    @ResponseBody
    public Result myFollowers(@RequestParam("userId")Long userId, @RequestParam("pageNum") int pageNum, @RequestParam("pageSize") int pageSize) {
        return Result.success(userService.myFollowers(userId, pageNum, pageSize));
    }

    @GetMapping("/test")
    @ResponseBody
    public Result test() {
        userService.addAuditPoint();
        return Result.success("aaaaa");
    }
}
